DSA PTP support

yangbolu1991 opened the pull request:
This PR is to support DSA PTP. (Now only gPTP is supported)
Also enabled i.MX943 gPTP sample for testing. gPTP on all 5 ports on i.MX943 (2 DSA switch ports and 3 ENETC ports) worked fine.

There is a hal_nxp dependency:
- https://github.com/zephyrproject-rtos/hal_nxp/pull/594

Thanks a lot.

yangbolu1991 committed:
dts: bindings: dsa: allow ptp-clock property

Allowed ptp-clock property for dsa.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: dsa: support tagger data and connecting device to tag protocol

Supported tagger data and connecting device to tag protocol, so that some
device specific work could be handled.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: dsa: move dsa_tag_netc.h to public include directory

Moved dsa_tag_netc.h to public include directory for devices
reusing.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: nxp_imx_netc: handle ports without PTP clock enabled

PTP clock may be enabled per Ethernet port. Especially pseudo port for
switch management does not need PTP clock.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: nxp_imx_netc: support receiving more types frames

Supported receiving more types frames. One was TX timestamp response
frame and the other one is RX timestamped frame which were for switch
port.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: dsa: add PTP solution in core driver

Added PTP solution in core driver. Now only gPTP was supported.

- ethernet_api like get_capabilities and get_ptp_clock
  were supported in port driver.

- For TX timestamp, added new dsa_api port_txtstamp for device
  to handle. Device driver might put pkt in queue, and reported
  timestamp after completing transmitting on hardware.

- For RX timestamp, the timestamp could be given to pkt in tag
  driver or device driver.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: pkt: support common usage of control block

Supported common usage of control block for any layer.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: gptp: add imx943_evk_mimx94398_m33_ddr support

Added imx943_evk_mimx94398_m33_ddr support.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
west.yml: update hal_nxp to support netc switch ptp

Updated hal_nxp to support netc switch ptp.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to fix twister issues.
Thanks.

ankuns reviewed (commented):
Reviewed from IEEE 802.15.4 point of view which as I guess is not your primary interest, however the change must not break anything there, should not make the code unsafe to use and should not lead to a waste of memory. Please have a look on the notes below.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -94,6 +94,9 @@ struct dsa_switch_context {
 	/** Number of initialized ports in the DSA switch */
 	uint8_t init_ports;
+
+	/** tagger data */
```
Will improve it. Thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -141,6 +152,10 @@ struct dsa_port_config {
 	const int tag_proto;
 	/** Ethernet device connected to the port */
 	const struct device *ethernet_connection;
+#ifdef CONFIG_NET_L2_PTP
+	/** ptp clock used on the port */
+	const struct device *ptp_clock;
+#endif
```
Will fix, thanks.

yangbolu1991 commented on subsys/net/ip/Kconfig:
```diff
@@ -755,6 +755,18 @@ config NET_PKT_BUF_TX_DATA_ALLOC_ALIGN_LEN
+	help
+	  This is to enable control block which can be used by any layers.
+
+config NET_PKT_CONTROL_BLOCK_SIZE
+	int "Size of pkt control block"
+	depends on NET_PKT_CONTROL_BLOCK
+	default 16
+	help
+	  Size of pkt control block.
```
Thanks a lot for the information. Now understand...
Will fix per your suggestion.

yangbolu1991 commented:
Updated to v2. Changes include,
- Improved API doc per @kartben suggestion.
- Converted to use union for pkt cb per @ankuns suggestion.

Thanks a lot.

ankuns reviewed (dismissed):
Looks OK from ieee802154 point of view, other parts not in my field of expertise.

yangbolu1991 commented:
Any other comments?
Thanks:)

kartben reviewed (dismissed):
+1 for docs, thanks!

yangbolu1991 committed:
drivers: ethernet: dsa_nxp_imx_netc: support PTP timestamping

Supported PTP timestamping. Now only gPTP was supported.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Rebased to address conflicts.

@jukkar  Could you help to review this PR?
Thank you very much.

jukkar reviewed (commented):
LGTM, I will add my +1 after the HAL is merged and manifest updated.

yangbolu1991 commented on subsys/net/l2/ethernet/dsa/dsa_core.c:
```diff
@@ -29,11 +29,40 @@ int dsa_xmit(const struct device *dev, struct net_pkt *pkt)
+	 * In case of using TX pkt in other places, pkt should not be changed.
+	 * Here just clone pkt to use for tagging and sending.
+	 */
+	clone = net_pkt_clone(pkt, K_NO_WAIT);
```
In zephyr, the origin pkt may be used by stack, for example, gPTP will use the origin pkt and its timestamp.
So, for dsa, we actually should not break origin pkt by tagging.
We have to clone pkt to use for tagging and sending.

yangbolu1991 commented on subsys/net/ip/Kconfig:
```diff
@@ -755,6 +755,21 @@ config NET_PKT_BUF_TX_DATA_ALLOC_ALIGN_LEN
+config NET_PKT_CONTROL_BLOCK
+	bool "pkt control block"
+	help
+	  This is to enable control block which can be used by any layers.
```
Thanks, will fix that.

yangbolu1991 commented on drivers/ethernet/dsa/dsa_nxp_imx_netc.c:
```diff
@@ -118,6 +163,70 @@ static void dsa_netc_port_phylink_change(const struct device *phydev, struct phy
+	}
+
+	pkt->cb.cb[0] = id;
+	id++;
```
The static id from 0 to 15 is used.

yangbolu1991 commented on drivers/ethernet/dsa/dsa_nxp_imx_netc.c:
```diff
@@ -118,6 +163,70 @@ static void dsa_netc_port_phylink_change(const struct device *phydev, struct phy
+	static uint8_t id = 1;
+
+	/* Utilize control block for timestamp request ID. */
+	if (id == 16) {
```
The static id from 0 to 15 is used.

yangbolu1991 commented on subsys/net/ip/Kconfig:
```diff
@@ -755,6 +755,21 @@ config NET_PKT_BUF_TX_DATA_ALLOC_ALIGN_LEN
+	  This is to enable control block which can be used by any layers.
+
+config NET_PKT_CONTROL_BLOCK_SIZE
+	int "Size of pkt control block"
```
Yes. Will add in description.

yangbolu1991 commented on subsys/net/l2/ethernet/dsa/dsa_tag_netc.c:
```diff
@@ -55,27 +99,53 @@ struct net_pkt *dsa_tag_netc_xmit(struct net_if *iface, struct net_pkt *pkt)
+#endif
+	tag_common->tpid = NETC_SWITCH_DEFAULT_ETHER_TYPE;
+	tag_common->type = kNETC_TagToPort;
+	tag_common->swtId = 1;
```
Yes. But we may consider to support multiple in the future.
Thanks.

yangbolu1991 reviewed (commented):
Thanks for the comments.
Sometimes I forgot to submit comments together in review. Sorry for that noise introduced by single comments.

yangbolu1991 commented:
Updated to v3. Changes include,
- Addressed west.yml conflict.
- Some code improvement per @JiafeiPan suggestin.

Thanks.

yangbolu1991 committed:
net: dsa: support PTP timestamping in netc tag driver

Supported PTP timestamping in netc tag driver.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>
